internal/controller/keystoneapplicationcredential_controller.go

```diff
@@ -0,0 +1,543 @@
+		var previousExpiresAt string
+		if isRotation && instance.Status.ExpiresAt != nil {
+			previousExpiresAt = instance.Status.ExpiresAt.Format(time.RFC3339)
+		}
+
+		// Update status
+		instance.Status.ACID = newID
+		instance.Status.SecretName = secretName
+		instance.Status.CreatedAt = &metav1.Time{Time: time.Now().UTC()}
+		instance.Status.ExpiresAt = &metav1.Time{Time: expiresAt}
+
+		// Calculate rotation eligible time (start of grace period window)
+		graceDuration := time.Duration(instance.Spec.GracePeriodDays) * 24 * time.Hour
+		rotationEligibleAt := expiresAt.Add(-graceDuration)
+		instance.Status.RotationEligibleAt = &metav1.Time{Time: rotationEligibleAt}
+
+		instance.Status.Conditions.MarkTrue(keystonev1.KeystoneApplicationCredentialReadyCondition, keystonev1.KeystoneApplicationCredentialReadyMessage)
+
+		// Set LastRotated and emit event if this was a rotation
+		if isRotation {
+			now := metav1.Now()
+			instance.Status.LastRotated = &now
+
+			// Emit event for EDPM visibility
+			r.EventRecorder.Event(
+				instance,
+				corev1.EventTypeNormal,
+				"ApplicationCredentialRotated",
+				fmt.Sprintf("ApplicationCredential '%s' (user: %s) rotated - EDPM nodes may need credential updates. Previous expiration: %s, New expiration: %s",
+					instance.Name, instance.Spec.UserName, previousExpiresAt, expiresAt.Format(time.RFC3339)),
+			)
+
+			logger.Info("ApplicationCredential rotated", "serviceName", instance.Spec.UserName)
+		} else {
+			logger.Info("ApplicationCredential created", "serviceName", instance.Spec.UserName)
+		}
+
+		// patch status immediately
+		if err := helperObj.PatchInstance(ctx, instance); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// requeue to check again before next rotate
+		nextCheck := r.computeNextRequeue(instance)
+		logger.Info("ApplicationCredential ready", "secret", secretName, "ACID", newID, "expiresAt", expiresAt, "nextCheck", nextCheck)
+		return ctrl.Result{RequeueAfter: nextCheck}, nil
+	}
+
+	// ApplicationCredential already exists and is valid
+	// Update RotationEligibleAt in case GracePeriodDays changed
+	if instance.Status.ExpiresAt != nil {
+		graceDuration := time.Duration(instance.Spec.GracePeriodDays) * 24 * time.Hour
+		rotationEligibleAt := instance.Status.ExpiresAt.Add(-graceDuration)
+		instance.Status.RotationEligibleAt = &metav1.Time{Time: rotationEligibleAt}
+	}
+
+	instance.Status.Conditions.MarkTrue(keystonev1.KeystoneApplicationCredentialReadyCondition, keystonev1.KeystoneApplicationCredentialReadyMessage)
+	nextCheck := r.computeNextRequeue(instance)
+	return ctrl.Result{RequeueAfter: nextCheck}, nil
+}
+
+func (r *ApplicationCredentialReconciler) reconcileDelete(
+	ctx context.Context,
+	instance *keystonev1.KeystoneApplicationCredential,
+) (ctrl.Result, error) {
+
+	// Before removing our CR finalizer, allow ApplicationCredential Secret to be deleted by removing its protection finalizer
```

Comment:
don't we first have to clean it up in keystone by doing an keystoneapi call, similar to how we do it in the keustoneendpoint https://github.com/openstack-k8s-operators/keystone-operator/blob/main/internal/controller/keystoneendpoint_controller.go#L255 ? otherwise we end up in the AC still be there in keystone? or is there some automatic cleanup in keystone and we just let it expire?